wnfs/src/public/file.rs

```diff
@@ -142,6 +125,24 @@ impl PublicFile {
     }
 }
 
+impl PrepareMut for PublicFile {
+    /// Takes care of creating previous links, in case the current
+    /// directory was previously `.store()`ed.
+    /// In any case it'll try to give you ownership of the directory if possible,
+    /// otherwise it clones.
+    fn prepare_mut<'a>(self: &'a mut Rc<Self>) -> &'a mut Self {
+        let Some(previous_cid) = self.persisted_as.get().cloned() else {
+            return Rc::make_mut(self);
+        };
+
+        let cloned = Rc::make_mut(self);
+        cloned.persisted_as = OnceCell::new();
+        cloned.previous = [previous_cid].into_iter().collect();
+
+        cloned
+    }
+}
+
```

Comment:
What is the reasoning behind introducing a trait? (I'm generally very careful about doing that, which is why I'm asking)
I don't see any function that is abstract over an `impl PrepareMut` or `P: PrepareMut` or similar.

Reply:
It is only there for reuse across the public and private side.

Reply:
Sorry, allow me to dig into this a little more:
Even if we had an `impl PrepareMut for PrivateDirectory` and `impl PrepareMut for PrivateFile`, etc.
we still wouldn't actually have any function that is abstract over "something that is `impl PrepareMut`". Everywhere that we would use `x.prepare_mut()`, we'd actually know the concrete type of `x`.
In these cases, the code would also work if we didn't use a trait for `.prepare_mut()`.

Reply:
It is just used to describe a common interface. It can be rid of; not that it matters too much.

Reply:
Okay. Yeah so we're somewhat on the same page. I just tend to avoid traits as much as possible. (This is me coming from haskell, where people have been "overusing" type classes a little too much, and which resulted in some people taking on [strong stances on using type classes](https://stackoverflow.com/a/17100055). Of course rust != haskell, but I think some insights transfer.)